modules/apps/frontend-js/frontend-js-state-web/src/main/resources/META-INF/resources/liferay/State.ts

```diff
@@ -0,0 +1,411 @@
+	}
+}
+
+const subscribers = new SubscriberMap();
+
+const warnings = {
+	readValue: new Set(),
+	writeValue: new Set(),
+};
+
+const State = {
+
+	// TODO: debug methods to read all atoms, all selectors etc.
+
+	atom<T>(key: string, value: T): Atom<T> {
+		if (atoms.has(key)) {
+			if (process.env.NODE_ENV === 'production') {
+
+				// Unlike in development where there may be hot-reloading,
+				// re-registering (overwriting) is not considered ok.
+
+				throw new Error(
+					`Liferay.State.atom(): key ${key} already taken`
+				);
+			}
+		}
+
+		if (selectors.has(key)) {
+			throw new Error(
+				`Liferay.State.atom(): key ${key} already taken by a selector`
+			);
+		}
+
+		const atom = deepFreeze({
+			default: value,
+			key,
+			[ATOM]: true,
+		} as const);
+
+		atoms.set(key, atom);
+
+		return atom;
+	},
+
+	// TODO: read() convenience method that takes an atom or selector
+	// TODO: (later, unsafe API taking a string key, for use in JSPs and such)
+
+	readAtom<T>(atom: Atom<T>): T {
+		if (values.has(atom)) {
+			return values.get(atom) as any;
+		}
+		else {
+			return atom.default;
+		}
+	},
+
+	readSelector<T>(selector: Selector<T>): T {
+		const seen = new Set<Selector<unknown>>();
+
+		return State._readSelector(selector, seen);
+	},
+
+	/*
+	readValue(key: string): unknown {
+		if (process.env.NODE_ENV === 'development') {
+			if (!warnings.readValue.has(key)) {
+				console.warn(
+					`Liferay.State.readValue(): access via string keys is not type-safe; prefer readAtom() or readSelector() instead`
+				);
+
+				warnings.readValue.add(key);
+			}
+		}
+
+		const atom = atoms.get(key);
+
+		if (atom) {
+			return State.readAtom(atom);
+		}
+	},
+	*/
+
+	selector<T>(key: string, getter: (get: Getter) => T) {
+		if (selectors.has(key)) {
+			if (process.env.NODE_ENV === 'production') {
+
+				// In development where there may be hot-reloading,
+				// re-registering (overwriting) is considered ok.
+
+				throw new Error(
+					`Liferay.State.selector(): key ${key} already taken`
+				);
+			}
+		}
+
+		if (atoms.has(key)) {
+			throw new Error(
+				`Liferay.State.selector(): key ${key} already taken by an atom`
+			);
+		}
+
+		const selector = deepFreeze({
+			key,
+			getter,
+			[SELECTOR]: true,
+		} as const);
+
+		selectors.set(key, selector);
+
+		return selector;
+	},
+
+	subscribe<T extends any>(
+		atomOrSelector: Atom<T> | Selector<T>,
+		callback: (value: T) => void
+	): {dispose: () => void} {
+		const dispose = subscribers.addCallback(atomOrSelector, callback);
+
+		if (!isAtom(atomOrSelector)) {
+
+			// Read once in order to pre-populate dependency graph.
+
+			State.readSelector(atomOrSelector);
+		}
+
+		return {dispose};
+	},
+
+	writeAtom<T>(atom: Atom<T>, value: T): void {
+		const previous = values.get(atom);
+
+		if (Object.is(value, previous)) {
+			return;
+		}
+
+		values.set(atom, value);
+
+		const callbacks = Array.from(subscribers.getCallbacks(atom).values());
```

Comment:
Note to self: remove the `Array.from()` here and lose the local variable. It is vestigial kruft from an intermediate state.